aesara/tensor/basic.py

```diff
@@ -1812,19 +1814,163 @@ def extract_constant(x, elemwise=True, only_process_constants=False):
     return x
 
 
-def transpose(x, axes=None):
-    """
-    Reorder the dimensions of x. (Default: reverse them)
+class TransposeOp(Op):
 
-    This is a macro around dimshuffle that matches the numpy.transpose function.
+    view_map = {0: [0]}
 
-    """
-    if axes is None:
-        axes = list(range((x.ndim - 1), -1, -1))
-    ret = DimShuffle(x.broadcastable, axes)(x)
-    if x.name and axes == list(range((x.ndim - 1), -1, -1)):
-        ret.name = x.name + ".T"
-    return ret
+    def make_node(self, input_arr, new_order=None):
+        input_arr = as_tensor_variable(input_arr)
+        ib = list(input_arr.type.broadcastable)
+
+        if new_order is None:
+            new_order = as_tensor_variable(np.flip(np.arange(input_arr.ndim)))
+            ib.reverse()
+            output = TensorType(dtype=input_arr.type.dtype, shape=tuple(ib))()
+            if input_arr.name:
+                output.name = input_arr.name + ".T"
+        else:
+            from aesara.tensor.var import TensorConstant
+
+            assert isinstance(new_order, TensorConstant)
+            # TODO: Try to remove this eval
+            _new_order = list(new_order.eval())
+            ob = []
+            if len(_new_order) != len(ib):
+                raise ValueError(
+                    f"Expected {len(ib)} number of values, found {len(_new_order)}."
+                )
+            for a in _new_order:
+                if a >= len(ib):
+                    raise ValueError("Found value greater than number of axes.")
+                if a < 0:
+                    raise ValueError(
+                        "Negative reference to axis, (e.g. axis = -1) is not allowed."
+                    )
+                ob.append(ib[a])
+            output = TensorType(dtype=input_arr.type.dtype, shape=ob)()
```

Comment:
If we force order to be a Constant, it seems like there's no good point in creating a new `Op`. The `len` can be obtained with `get_vector_length`. The output shape might need something more lax, if known (via e.g., `get_constant_value`), set to previous shape dimension, otherwise set to `None` (unknown)